### PR TITLE
Add HttpLongRunningOperationResponse

### PR DESCRIPTION
**Pull request by daschult**
Resolves https://github.com/Azure/azure-sdk-for-node/issues/2871 (partially).

This PR adds the HttpLongRunningOperationResponse type to ms-rest-azure-js. This is the public facing polling object that encapsulates the complexity of the LROPollStrategy types and exposes a nicer facade that customers can work with. It enables customers to get the current status of their LRO, to do a single poll, to poll until the LRO is finished, or to get a memento of the LRO that can be saved to disk or sent across the network. This memento can later be used to restore the HttpLongRunningOperationStatus and continue polling.

### CODE REVIEW

File: lib/httpLongRunningOperationResponse.ts [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+   * different process or a different machine). If the LRO couldn't produce an LRO polling strategy,
+   * then this will return undefined.
+   */
+  public getMemento(): LROMemento | undefined {
+    const lroPollStrategy: LROPollStrategy | undefined = this._lroPollStrategy;
+    return !lroPollStrategy ? undefined : lroPollStrategy.getMemento();
+  }
+}
+
+export function createHttpLongRunningOperationResponseFromInitialResponse(azureServiceClient: AzureServiceClient, initialResponse: HttpOperationResponse, options?: RequestOptionsBase): HttpLongRunningOperationResponse {
```
Comment (RikkiGibson) [inline lib/httpLongRunningOperationResponse.ts]: What makes `export` necessary for these 2 functions?

File: lib/lroPollStrategy.ts [hunk 2]
```diff
@@ -230,21 +245,61 @@ function isFinished(status: LongRunningOperationStates): boolean {
...
+      mostRecentRequest: initialResponse.request,
+      azureAsyncOperationHeaderValue: getAzureAsyncOperationHeaderValue(initialResponse),
+      locationHeaderValue: getLocationHeaderValue(initialResponse),
+      resource: resource,
+      state: getStatusFromResponse(initialResponse, resource)
+    };
+    result = createLROPollStrategyFromMemento(azureServiceClient, lroMemento);
   } else {
-    throw new Error("Can't determine long running operation polling strategy from initial response.");
+    result = undefined;
```
Comment (RikkiGibson) [inline lib/lroPollStrategy.ts]: Does it ever make sense to hit this line?

File: lib/httpLongRunningOperationResponse.ts (hunk 1 above)
Comment (daschult) [reply lib/httpLongRunningOperationResponse.ts]: These two functions serve as the constructor functions for HttpLongRunningOperationResponse objects. They're both used by the AzureServiceClient class.

File: lib/lroPollStrategy.ts (hunk 2 above)
Comment (daschult) [reply lib/lroPollStrategy.ts]: This line shouldn't ever be hit if all of our services properly follow the LRO specifications, but if they do something that is unexpected or if the initial response indicates that the operation has already completed, then this line will be hit.

**Comment by RikkiGibson** [comment]
Is there any need to bump the minor version before publishing?

**Comment by daschult** [comment]
Not yet. There weren't any breaking changes in this PR. Just additive changes.

**Pull request merged**